fairlearn/grid_search.py

```diff
@@ -0,0 +1,127 @@
+        # Nothing to do
+        return
+
+    def _generate_p0_p1(self, protected_attribute):
+        """ Function to compute p0 and p1 for the given
+        set of protected_attribute values
+        """
+        unique_labels, counts = np.unique(protected_attribute, return_counts=True)
+        if len(unique_labels) > 2:
+            raise RuntimeError("Protected Attribute is not binary")
+
+        p0 = counts[0] / len(protected_attribute)
+        # It is possible that our data only contain one value for
+        # the protected attribute
+        # We do not consider this an error
+        if len(counts) > 1:
+            p1 = counts[1] / len(protected_attribute)
+        else:
+            p1 = 0
+
+        return p0, p1
+
+    def _weight_function(self, y_val, a_val, L, p_ratio):
+        if a_val == 0:
+            return 2*y_val - 1 - L * p_ratio
+        else:
+            return 2*y_val - 1 + L
+
+    def _generate_weights(self, y, protected_attribute, L, p_ratio):
+        weight_func = np.vectorize(self._weight_function)
+        return weight_func(y, protected_attribute, L, p_ratio)
+
+    def demographic_parity_binary_classification(self, learner, x, y, protected_attribute, lagrange_multipliers=None, number_lagrange_multipliers=11):
+        """Function to generate a list of models for a binary classification problem with
+        a single binary protected attribute. The models sweep through different potential
+        lagrange multipliers for the constrained optimisation problem (the constraint
+        being demographic parity), each one corresponding to a particular tradeoff between
+        fairness and accuracy
+
+        :param learner: An object which can be used to fit a model to features, labels and
+        weights. A deep copy of this is made for each value of the lagrange multiplier used
+        :type learner: Must implement a fit(x, y, sample_weight) method
+
+        :param x: The array of training data features (which may or may not contain the
+        protected attribute). Must have as many rows as y and protected_attribute
+        :type x: Nested list or numpy array with two dimensions or pandas dataframe
+
+        :param y: The list of binary classes, which must be 0 or 1. Must contain the same
+        number of entries as rows in x. We do not consider it an error if only one
+        class is present
+        :type y: List or Numpy array with one dimension
+
+        :param protected_attribute: The binary protected attribute corresponding to each row of x
+        We do not consider it an error if only one attribute value is present
+        :type protected_attribute: List of Numpy array with one dimension
+        
+        :param lagrange_multipliers: User specified set of lagrange multipliers to use for
+        the optimisation problem. If this is set then num_Ls must be None. The result array will
+        be equal in length to this array
+        :type lagrange_multipliers: List of real numbers
```

Comment:
I think people use this with actual python types, so in this case 
```
:type lagrange_multipliers: list
```

Reply:
What do they do when multiple types are possible? For example A, Y and X can be a variety of different things

Reply:
Hmm good question. Maybe it's best to just leave it as is then...